src/psyclone/domain/gocean/transformations/kern_trans.py

```diff
@@ -0,0 +1,163 @@
+                break
+            except KeyError:
+                pass
+        else:
+            raise TransformationError(
+                f"The metadata name ({self._metadata_name}) provided to the "
+                f"transformation does not correspond to a symbol in the "
+                f"supplied PSyIR.")
+
+        # Validate the metadata. This is done as part of the setup()
+        # method. This method can't be called until we specialise the
+        # symbol, so we specialise with a copy of the symbol.
+        tmp_metadata_symbol = metadata_symbol.copy()
+        tmp_metadata_symbol.specialise(KernelMetadataSymbol)
+        tmp_metadata_symbol.setup()
+
+        if not isinstance(node, Container):
+            raise TransformationError(
+                f"Error in {self.name} transformation. The supplied call "
+                f"argument should be a Container node but found "
+                f"'{type(node).__name__}'.")
+
+        if node.parent:
+            raise TransformationError(
+                f"Error in {self.name} transformation. The supplied node "
+                f"should be the root of a PSyIR tree but this node has a "
+                f"parent.")
+
+    @property
+    def metadata_name(self):
+        '''
+        :returns: the name of the metadata that determines what will be \
+            specialised to kernel-specific PSyIR.
+        :rtype: str
+
+        '''
+        return self._metadata_name
+
+    @metadata_name.setter
+    def metadata_name(self, value):
+        '''
+        :param str value: sets the name of the metadata that \
+            determines what will be specialised to kernel-specific PSyIR.
+        '''
+        if not isinstance(value, str):
+            raise TypeError(
+                f"The kern_trans transformation requires the metadata name "
+                f"to be a string, but found '{type(value).__name__}'.")
+
+        self._metadata_name = value
+
+    def apply(self, node, options=None):
+        ''' Apply transformation to the supplied PSyIR.
+
+        :param node: a kernel represented in generic PSyIR.
+        :type node: Union[:py:class:`psyclone.psyir.node.Routine`, \
+            :py:class:`psyclone.psyir.node.Container`]
+        :param options: a dictionary with options for transformations.
+        :type options: Optional[Dict[str: str]]
+
+        '''
+        self.validate(node, options=options)
+
+        # Find the metadata symbol. No need to check it is found as
+        # this is done in the validate method.
+        metadata_symbol = None
+        for schedule_node in node.walk(Schedule):
+            try:
+                metadata_symbol = schedule_node.symbol_table.lookup(
+                    self._metadata_name)
+                break
+            except KeyError:
+                pass
+
+        # Transform the kernel metadata
+        metadata_symbol.specialise(KernelMetadataSymbol)
+        metadata_symbol.setup()
```

Comment:
I'm not sure about this. It seems that the metadata symbol is transforming itself. I would have expected this transformation to take the information from the PSyIR and use it to construct the metadata symbol (by passing the distilled information to a constructor/create method). Perhaps this comes back to my concerns about the fact that the metadata symbol class setters update two quantities instead of just one?

Reply:
This has all been reworked after discussion.